pkg/endpoint/endpoint_status.go

```diff
@@ -0,0 +1,228 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package endpoint
+
+import (
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/cilium/cilium/api/v1/models"
+	"github.com/cilium/cilium/pkg/identity"
+	identitycache "github.com/cilium/cilium/pkg/identity/cache"
+	cilium_v2 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2"
+	"github.com/cilium/cilium/pkg/labels"
+	"github.com/cilium/cilium/pkg/policy/trafficdirection"
+)
+
+func getEndpointStatusControllers(status *models.EndpointStatus) (controllers cilium_v2.ControllerList) {
+	for _, c := range status.Controllers {
+		if c.Status == nil {
+			continue
+		}
+
+		if c.Status.ConsecutiveFailureCount > 0 {
+			s := cilium_v2.ControllerStatus{
+				Configuration: c.Configuration,
+				Name:          c.Name,
+				UUID:          string(c.UUID),
+				Status: cilium_v2.ControllerStatusStatus{
+					ConsecutiveFailureCount: c.Status.ConsecutiveFailureCount,
+					FailureCount:            c.Status.FailureCount,
+					LastFailureMsg:          c.Status.LastFailureMsg,
+					LastFailureTimestamp:    c.Status.LastFailureTimestamp.String(),
+					LastSuccessTimestamp:    c.Status.LastSuccessTimestamp.String(),
+					SuccessCount:            c.Status.SuccessCount,
+				},
+			}
+			if controllers == nil {
+				controllers = cilium_v2.ControllerList{s}
+			} else {
+				controllers = append(controllers, s)
+			}
+		}
+
+		if controllers != nil {
+			controllers.Sort()
+		}
+	}
+	return
+}
+
+func getEndpointStatusLog(e *Endpoint) (log []*models.EndpointStatusChange) {
+	added := 0
+
+	if s := e.Status; s != nil {
+		s.indexMU.RLock()
+		defer s.indexMU.RUnlock()
+
+		for i := s.lastIndex(); ; i-- {
+			if i < 0 {
+				i = maxLogs - 1
+			}
+			if i < len(s.Log) && s.Log[i] != nil {
+				l := &models.EndpointStatusChange{
+					Timestamp: s.Log[i].Timestamp.Format(time.RFC3339),
+					Code:      s.Log[i].Status.Code.String(),
+					Message:   s.Log[i].Status.Msg,
+					State:     models.EndpointState(s.Log[i].Status.State),
+				}
+
+				if strings.ToLower(l.Code) != models.EndpointStatusChangeCodeOk {
+					if log == nil {
+						log = []*models.EndpointStatusChange{l}
+					} else {
+						log = append(log, l)
+					}
+					added++
+					if added >= cilium_v2.EndpointStatusLogEntries {
+						break
+					}
+				}
+			}
+			if i == s.Index {
+				break
+			}
+		}
+	}
+	return
+}
+
+func getEndpointIdentity(status *models.EndpointStatus) (identity *cilium_v2.EndpointIdentity) {
+	if status.Identity != nil {
+		identity = &cilium_v2.EndpointIdentity{
+			ID:     status.Identity.ID,
+			Labels: status.Identity.Labels,
+		}
+
+		sort.Strings(identity.Labels)
```

Comment:
This line is modifying the internal array of `status.Identity.Labels`. Is it guaranteed that `status.Identity.Labels` will not be modified concurrently?